### PR TITLE
Day picker should dispose created MaterialStatesController's.

### LINKED ISSUE #133862: _DayPickerState should not create MaterialStatesController on every build, without disposing
PR: https://github.com/flutter/flutter/pull/133863

### PR DESCRIPTION
**Pull request by polina-c**
Fixes https://github.com/flutter/flutter/issues/133862

### CODE REVIEW

File: packages/flutter/lib/src/material/calendar_date_picker.dart
Comment (chunhtai) [inline packages/flutter/lib/src/material/calendar_date_picker.dart:1016]: This doesn't look right, every date widget should have a separate states controller. cc @HansMuller

Comment (HansMuller) [reply packages/flutter/lib/src/material/calendar_date_picker.dart:1016]: That's right. each day widget should have its own selected/disabled/etc state.

Comment (polina-c) [reply packages/flutter/lib/src/material/calendar_date_picker.dart:1016]: It is interesting that tests passed...

Comment (polina-c) [reply packages/flutter/lib/src/material/calendar_date_picker.dart]: Fixed. Looks better?

File: packages/flutter/lib/src/material/calendar_date_picker.dart
Comment (chunhtai) [inline packages/flutter/lib/src/material/calendar_date_picker.dart:1019]: This may leak if this widget rebuild with different days.

One idea to make this easier to manage is to create a statefulwidget wrapper around the dayWidget that holds the statesController. that way you won't need to keep track of each statesController in this widget.

File: packages/flutter/lib/src/material/calendar_date_picker.dart
Comment (HansMuller) [reply packages/flutter/lib/src/material/calendar_date_picker.dart]: Yes that looks better although factoring out a stateful day widget per @chunhtai suggestion seems like a cleaner way to do this. 

And, [as you pointed out](https://github.com/flutter/flutter/pull/133884#discussion_r1319168307), it looks like need some additional tests.

File: packages/flutter/lib/src/material/calendar_date_picker.dart
Comment (polina-c) [reply packages/flutter/lib/src/material/calendar_date_picker.dart:1019]: It will not leak because the map's keys are from 1 to 31.
So' it will create maximum 31 controllers. Even if month changes, controllers for days will get new correct values.
Anything I am missing?

Comment (chunhtai) [reply packages/flutter/lib/src/material/calendar_date_picker.dart:1019]: yes you are right that it will not grow forever. I am fine with this change then. But since @HansMuller is more familiar with this part of the code, we should get his LGTM before merge

**Review by chunhtai — APPROVED** [review]
LGTM, but it is better to wait for @HansMuller's approval

**Comment by HansMuller** [comment]
I'm OK with the change as it is although I'd prefer to see a Day widget factored out and a regression test that would fail if didn't properly maintain state per day (https://github.com/flutter/flutter/pull/133884#discussion_r1319182369).

File: packages/flutter/lib/src/material/calendar_date_picker.dart
Comment (polina-c) [reply packages/flutter/lib/src/material/calendar_date_picker.dart]: Added test.

I expect creation of separate widget for a day will take 2-3 more iterations, while I want to use the time to test cover other tests for leaks. 
Added TODO about it.

Comment (polina-c) [reply packages/flutter/lib/src/material/calendar_date_picker.dart]: I expect separate widget for a day will take 2-3 iterations, while I prefer to prioritize test coverage. Created TODO: https://github.com/flutter/flutter/issues/134323.

**Pull request merged**
